### ISSUE TITLE
Create all file copies simulatenously

### ISSUE THREAD

**Issue by gboudreau** [gboudreau/Greyhole#94]
Instead of executing three `cp` commands, execute just one `cat | tee` command, as follow:

```
cat <inputfile> | tee <outfile1> <outfile2> <outfile3> > /dev/null
```

This should speed up a lot the handling of new or modified file copies on multiple-copies shares.

**Comment by sabatech** [comment]
I was thinking about this myself, and in theory it sounds good, but what if a storage pool has a slow NFS volume or maybe usb2 volume attached... I am wondering how it is going to act when one <outfile> writes SEVERELY slower than the others. It's a potential problem, but who knows, it might be nothing at all.

**Comment by gboudreau** [comment]
Tested by creating/modifying a 1GB file on a 8-copies share:
- Before: 2m 11s
- After: 44s

There's also a lot less logs too; 113 lines before; 32 after.

On a 2-copies share, the improvement is a lot more modest:
- Before: 24s
- After: 21s

Before:
```
Mar 15 20:22:20 DEBUG write:   Destination /mnt/hdd5/gh/Photos/test_file_1GB is the same as the source. Nothing to do here; this file copy is ready!
Mar 15 20:22:20 DEBUG write:   Creating symlink in share pointing to /mnt/hdd5/gh/Photos/test_file_1GB
Mar 15 20:22:20 DEBUG write:   Saving 8 metadata files for Photos/test_file_1GB
Mar 15 20:22:20 DEBUG write:   Copying 1,024MB file to /mnt/hdd9/gh/Photos/test_file_1GB
Mar 15 20:22:37 DEBUG write:     Copied file MD5 = cd573cfaace07e7949bc0c46028904ff
Mar 15 20:22:37 DEBUG write:     Copy created at 60.2 MBps.
Mar 15 20:22:37 DEBUG write:   Saving 8 metadata files for Photos/test_file_1GB
Mar 15 20:22:37 DEBUG write:   Copying 1,024MB file to /mnt/hdd1/gh/Photos/test_file_1GB
Mar 15 20:22:55 DEBUG write:     Copied file MD5 = cd573cfaace07e7949bc0c46028904ff
Mar 15 20:22:55 DEBUG write:     Copy created at 56.9 MBps.
Mar 15 20:22:55 DEBUG write:   Saving 8 metadata files for Photos/test_file_1GB
Mar 15 20:22:55 DEBUG write:   Copying 1,024MB file to /mnt/hdd4/gh/Photos/test_file_1GB
Mar 15 20:23:13 DEBUG write:     Copied file MD5 = cd573cfaace07e7949bc0c46028904ff
Mar 15 20:23:13 DEBUG write:     Copy created at 56.9 MBps.
Mar 15 20:23:13 DEBUG write:   Saving 8 metadata files for Photos/test_file_1GB
Mar 15 20:23:13 DEBUG write:   Copying 1,024MB file to /mnt/hdd8/gh/Photos/test_file_1GB
Mar 15 20:23:31 DEBUG write:     Copied file MD5 = cd573cfaace07e7949bc0c46028904ff
Mar 15 20:23:31 DEBUG write:     Copy created at 56.9 MBps.
Mar 15 20:23:31 DEBUG write:   Saving 8 metadata files for Photos/test_file_1GB
Mar 15 20:23:31 DEBUG write:   Copying 1,024MB file to /mnt/hdd2/gh/Photos/test_file_1GB
Mar 15 20:23:56 DEBUG write:     Copied file MD5 = cd573cfaace07e7949bc0c46028904ff
Mar 15 20:23:56 DEBUG write:     Copy created at 41.0 MBps.
Mar 15 20:23:56 DEBUG write:   Saving 8 metadata files for Photos/test_file_1GB
Mar 15 20:23:56 DEBUG write:   Copying 1,024MB file to /mnt/hdd3/gh/Photos/test_file_1GB
Mar 15 20:24:14 DEBUG write:     Copied file MD5 = cd573cfaace07e7949bc0c46028904ff
Mar 15 20:24:14 DEBUG write:     Copy created at 56.9 MBps.
Mar 15 20:24:14 DEBUG write:   Saving 8 metadata files for Photos/test_file_1GB
Mar 15 20:24:14 DEBUG write:   Copying 1,024MB file to /mnt/hdd6/gh/Photos/test_file_1GB
Mar 15 20:24:30 DEBUG write:     Copied file MD5 = cd573cfaace07e7949bc0c46028904ff
Mar 15 20:24:30 DEBUG write:     Copy created at 64.0 MBps.
Mar 15 20:24:30 DEBUG write:   Saving 8 metadata files for Photos/test_file_1GB
```

After:
```
Mar 15 20:26:50 DEBUG write:   Destination /mnt/hdd5/gh/Photos/test_file_1GB is the same as the source. Nothing to do here; this file copy is ready!
Mar 15 20:26:50 DEBUG write:   Copying 1,024MB file to /mnt/hdd9/gh/Photos/test_file_1GB, /mnt/hdd1/gh/Photos/test_file_1GB, /mnt/hdd4/gh/Photos/test_file_1GB, /mnt/hdd8/gh/Photos/test_file_1GB, /mnt/hdd2/gh/Photos/test_file_1GB, /mnt/hdd3/gh/Photos/test_file_1GB, /mnt/hdd6/gh/Photos/test_file_1GB
Mar 15 20:27:33 DEBUG write:     Copied file MD5 = cd573cfaace07e7949bc0c46028904ff
Mar 15 20:27:33 DEBUG write:     Copy created at 23.8 MBps.
Mar 15 20:27:33 DEBUG write:   Creating symlink in share pointing to /mnt/hdd5/gh/Photos/test_file_1GB
Mar 15 20:27:33 DEBUG write:   Saving 8 metadata files for Photos/test_file_1GB
```

Not sure why parallel copying is much slower than even the slowest of HDD; maybe it has to do with CPU io-wait... Will do more tests.

**Comment by gboudreau** [comment]
For a 12GB file, sequential copying (before): 5m 52s
```
Mar 18 18:35:12 DEBUG write:   Copying 12.0GB file to /mnt/hdd3/gh/Videos/test-file-12gb-2
Mar 18 18:37:30 DEBUG write:     Copied file MD5 = 4ae38594bd7793ec57aaf5643372ac37
Mar 18 18:37:30 DEBUG write:     Copy created at 89.0 MBps.
Mar 18 18:37:30 DEBUG write:   Creating symlink in share pointing to /mnt/hdd3/gh/Videos/test-file-12gb-2
Mar 18 18:37:30 DEBUG write:   Saving 2 metadata files for Videos/test-file-12gb-2
Mar 18 18:37:30 DEBUG write:   Copying 12.0GB file to /mnt/hdd6/gh/Videos/test-file-12gb-2
Mar 18 18:41:04 DEBUG write:     Copied file MD5 = 4ae38594bd7793ec57aaf5643372ac37
Mar 18 18:41:04 DEBUG write:     Copy created at 57.4 MBps.
Mar 18 18:41:04 DEBUG write:   Saving 2 metadata files for Videos/test-file-12gb-2
```

Parallel copying (after): 3m 30s
```
Mar 18 18:02:24 DEBUG write:   Copying 12.0GB file to: /mnt/hdd3/gh/Videos/.test-file-12gb.15aa1, /mnt/hdd6/gh/Videos/.test-file-12gb.15aa1
Mar 18 18:05:54 DEBUG write:     Copied file MD5 = 4ae38594bd7793ec57aaf5643372ac37
Mar 18 18:05:54 DEBUG write:     Copy created at 58.5 MBps.
Mar 18 18:05:54 DEBUG write:   Creating symlink in share pointing to /mnt/hdd3/gh/Videos/test-file-12gb
Mar 18 18:05:54 DEBUG write:   Saving 2 metadata files for Videos/test-file-12gb
```

**Comment by assafw** [comment]
Hi,

Recently updated and this is indeed faster when the source and target copies are not on the same drives. but, if the LZ is configured on one of the pool drives, copies are very slow (read and write to the same physical drive at the same time), and 'mv' is faster.

Is there a way to revert to the previous behavior (through a configuration switch)?

Cheers!

**Comment by gboudreau** [comment]
@assafw Please open another issue, and include your system information: https://github.com/gboudreau/Greyhole/wiki/GetHelp#gather-system-information

**Comment by assafw** [comment]
Hello,

Really appreciate the work you do! I've been using gh for a while and
generally very happy with it!

Since my previous comment was deleted, I think it had something to do with
My LZ on the same drive as one of the data drives, so copies are slow. In
general I found that on my machine cat | tee f1 f2 is much slower and avg
copy speed is ~24MB/s, where as sequential copies were ~90-100mb/s

My point was that this needs to be an option, even just for the sake of
backward compatibility. greyhole does not have an easy path of
roll-backwards. I tried to install 0.11-2 on my machine without success
(and I can't apt pin it as it's no longer available from the apt repo).

Thoughts?
